File: internal/repository/sequence_repo.go
package repository

import (
	"context"
	"database/sql"
	"fmt"
	"time"

	"github.com/zajca/zfaktury/internal/domain"
)

// SequenceRepository handles persistence of InvoiceSequence entities.
type SequenceRepository struct {
	db *sql.DB
}

// NewSequenceRepository creates a new SequenceRepository.
func NewSequenceRepository(db *sql.DB) *SequenceRepository {
	return &SequenceRepository{db: db}
}

// Create inserts a new invoice sequence into the database.
func (r *SequenceRepository) Create(ctx context.Context, seq *domain.InvoiceSequence) error {
	now := time.Now()

	result, err := r.db.ExecContext(ctx, `
		INSERT INTO invoice_sequences (prefix, next_number, year, format_pattern, created_at, updated_at)
		VALUES (?, ?, ?, ?, ?, ?)`,
		seq.Prefix, seq.NextNumber, seq.Year, seq.FormatPattern, now, now,
	)
	if err != nil {
		return fmt.Errorf("inserting invoice sequence: %w", err)
	}

	id, err := result.LastInsertId()
	if err != nil {
		return fmt.Errorf("getting last insert id for invoice sequence: %w", err)
	}
	seq.ID = id
	return nil
}

// Update modifies an existing invoice sequence.
func (r *SequenceRepository) Update(ctx context.Context, seq *domain.InvoiceSequence) error {
	now := time.Now()

	result, err := r.db.ExecContext(ctx, `
		UPDATE invoice_sequences SET
			prefix = ?, next_number = ?, year = ?, format_pattern = ?, updated_at = ?
		WHERE id = ? AND deleted_at IS NULL`,
		seq.Prefix, seq.NextNumber, seq.Year, seq.FormatPattern, now, seq.ID,
	)
	if err != nil {
		return fmt.Errorf("updating invoice sequence %d: %w", seq.ID, err)
	}

	rows, err := result.RowsAffected()
	if err != nil {
		return fmt.Errorf("checking rows affected for invoice sequence %d: %w", seq.ID, err)
	}
	if rows == 0 {
		return fmt.Errorf("invoice sequence %d not found or already deleted", seq.ID)
	}
	return nil
}

// Delete performs a soft delete on an invoice sequence.
func (r *SequenceRepository) Delete(ctx context.Context, id int64) error {
	now := time.Now()
	result, err := r.db.ExecContext(ctx, `
		UPDATE invoice_sequences SET deleted_at = ?, updated_at = ? WHERE id = ? AND deleted_at IS NULL`,
		now, now, id,
	)
	if err != nil {
		return fmt.Errorf("soft-deleting invoice sequence %d: %w", id, err)
	}

	rows, err := result.RowsAffected()
	if err != nil {
		return fmt.Errorf("checking rows affected for invoice sequence %d: %w", id, err)
	}
	if rows == 0 {
		return fmt.Errorf("invoice sequence %d not found or already deleted", id)
	}
	return nil
}

// GetByID retrieves a single invoice sequence by ID.
func (r *SequenceRepository) GetByID(ctx context.Context, id int64) (*domain.InvoiceSequence, error) {
	seq := &domain.InvoiceSequence{}

	err := r.db.QueryRowContext(ctx, `
		SELECT id, prefix, next_number, year, format_pattern
		FROM invoice_sequences WHERE id = ? AND deleted_at IS NULL`, id,
	).Scan(&seq.ID, &seq.Prefix, &seq.NextNumber, &seq.Year, &seq.FormatPattern)
	if err != nil {
		if err == sql.ErrNoRows {
			return nil, fmt.Errorf("invoice sequence %d not found: %w", id, err)
		}
		return nil, fmt.Errorf("querying invoice sequence %d: %w", id, err)
	}
	return seq, nil
}

// List retrieves all non-deleted invoice sequences.
func (r *SequenceRepository) List(ctx context.Context) ([]domain.InvoiceSequence, error) {
	rows, err := r.db.QueryContext(ctx, `
		SELECT id, prefix, next_number, year, format_pattern
		FROM invoice_sequences
		WHERE deleted_at IS NULL
		ORDER BY year DESC, prefix ASC`)
	if err != nil {
		return nil, fmt.Errorf("listing invoice sequences: %w", err)
	}
	defer rows.Close()

	var sequences []domain.InvoiceSequence
	for rows.Next() {
		var seq domain.InvoiceSequence
		if err := rows.Scan(&seq.ID, &seq.Prefix, &seq.NextNumber, &seq.Year, &seq.FormatPattern); err != nil {
			return nil, fmt.Errorf("scanning invoice sequence row: %w", err)
		}
		sequences = append(sequences, seq)
	}
	if err := rows.Err(); err != nil {
		return nil, fmt.Errorf("iterating invoice sequence rows: %w", err)
	}
	return sequences, nil
}

// GetByPrefixAndYear retrieves an invoice sequence by its prefix and year combination.
func (r *SequenceRepository) GetByPrefixAndYear(ctx context.Context, prefix string, year int) (*domain.InvoiceSequence, error) {
	seq := &domain.InvoiceSequence{}

	err := r.db.QueryRowContext(ctx, `
		SELECT id, prefix, next_number, year, format_pattern
		FROM invoice_sequences
		WHERE prefix = ? AND year = ? AND deleted_at IS NULL`, prefix, year,
	).Scan(&seq.ID, &seq.Prefix, &seq.NextNumber, &seq.Year, &seq.FormatPattern)
	if err != nil {
		if err == sql.ErrNoRows {
			return nil, fmt.Errorf("invoice sequence with prefix %q and year %d not found: %w", prefix, year, err)
		}
		return nil, fmt.Errorf("querying invoice sequence by prefix %q and year %d: %w", prefix, year, err)
	}
	return seq, nil
}

// CountInvoicesBySequenceID returns the number of invoices referencing a given sequence.
func (r *SequenceRepository) CountInvoicesBySequenceID(ctx context.Context, sequenceID int64) (int, error) {
	var count int
	err := r.db.QueryRowContext(ctx, `
		SELECT COUNT(*) FROM invoices WHERE sequence_id = ? AND deleted_at IS NULL`, sequenceID,
	).Scan(&count)
	if err != nil {
		return 0, fmt.Errorf("counting invoices for sequence %d: %w", sequenceID, err)
	}
	return count, nil
}

// MaxUsedNumber returns the highest invoice number used for a given sequence,
// by looking at the invoices table. Returns 0 if no invoices reference this sequence.
func (r *SequenceRepository) MaxUsedNumber(ctx context.Context, sequenceID int64) (int, error) {
	// We need to figure out the highest number used. Since invoice_number is formatted,
	// we count the number of invoices and use the current next_number - 1 as proxy.
	// A more reliable approach: next_number - 1 is the last assigned number.
	var nextNumber int
	err := r.db.QueryRowContext(ctx, `
		SELECT next_number FROM invoice_sequences WHERE id = ? AND deleted_at IS NULL`, sequenceID,
	).Scan(&nextNumber)
	if err != nil {
		if err == sql.ErrNoRows {
			return 0, nil
		}
		return 0, fmt.Errorf("querying max used number for sequence %d: %w", sequenceID, err)
	}
	// The last assigned number is next_number - 1 (since next_number is what will be assigned next).
	if nextNumber <= 1 {
		return 0, nil
	}
	return nextNumber - 1, nil
}

File: rust/crates/zfaktury-db/src/repos/sequence_repo.rs
use std::sync::Mutex;

use chrono::Local;
use rusqlite::{Connection, Row, params};
use zfaktury_core::repository::traits::InvoiceSequenceRepo;
use zfaktury_domain::{DomainError, InvoiceSequence};

use crate::helpers::*;

pub struct SqliteSequenceRepo {
    conn: Mutex<Connection>,
}

impl SqliteSequenceRepo {
    pub fn new(conn: Connection) -> Self {
        Self {
            conn: Mutex::new(conn),
        }
    }
}

fn scan_sequence(row: &Row<'_>) -> rusqlite::Result<InvoiceSequence> {
    Ok(InvoiceSequence {
        id: row.get("id")?,
        prefix: row.get("prefix")?,
        next_number: row.get("next_number")?,
        year: row.get("year")?,
        format_pattern: row.get("format_pattern")?,
    })
}

impl InvoiceSequenceRepo for SqliteSequenceRepo {
    fn create(&self, seq: &mut InvoiceSequence) -> Result<(), DomainError> {
        let conn = self.conn.lock().unwrap();
        conn.execute(
            "INSERT INTO invoice_sequences (prefix, next_number, year, format_pattern) VALUES (?1, ?2, ?3, ?4)",
            params![seq.prefix, seq.next_number, seq.year, seq.format_pattern],
        ).map_err(|e| { log::error!("inserting sequence: {e}"); DomainError::InvalidInput })?;
        seq.id = conn.last_insert_rowid();
        Ok(())
    }

    fn update(&self, seq: &mut InvoiceSequence) -> Result<(), DomainError> {
        let conn = self.conn.lock().unwrap();
        let now = format_datetime(&Local::now().naive_local());
        conn.execute(
            "UPDATE invoice_sequences SET prefix = ?1, next_number = ?2, year = ?3, format_pattern = ?4, updated_at = ?5 WHERE id = ?6 AND deleted_at IS NULL",
            params![seq.prefix, seq.next_number, seq.year, seq.format_pattern, now, seq.id],
        ).map_err(|e| { log::error!("updating sequence: {e}"); DomainError::InvalidInput })?;
        Ok(())
    }

    fn delete(&self, id: i64) -> Result<(), DomainError> {
        let conn = self.conn.lock().unwrap();
        let now = format_datetime(&Local::now().naive_local());
        let rows = conn.execute(
            "UPDATE invoice_sequences SET deleted_at = ?1, updated_at = ?1 WHERE id = ?2 AND deleted_at IS NULL",
            params![now, id],
        ).map_err(|e| { log::error!("deleting sequence: {e}"); DomainError::InvalidInput })?;
        if rows == 0 {
            return Err(DomainError::NotFound);
        }
        Ok(())
    }

    fn get_by_id(&self, id: i64) -> Result<InvoiceSequence, DomainError> {
        let conn = self.conn.lock().unwrap();
        conn.query_row(
            "SELECT id, prefix, next_number, year, format_pattern FROM invoice_sequences WHERE id = ?1 AND deleted_at IS NULL",
            params![id], scan_sequence,
        ).map_err(|e| match e {
            rusqlite::Error::QueryReturnedNoRows => DomainError::NotFound,
            _ => { log::error!("querying sequence: {e}"); DomainError::InvalidInput }
        })
    }

    fn list(&self) -> Result<Vec<InvoiceSequence>, DomainError> {
        let conn = self.conn.lock().unwrap();
        let mut stmt = conn.prepare("SELECT id, prefix, next_number, year, format_pattern FROM invoice_sequences WHERE deleted_at IS NULL ORDER BY year DESC, prefix ASC")
            .map_err(|e| { log::error!("preparing sequence list: {e}"); DomainError::InvalidInput })?;
        let rows = stmt.query_map([], scan_sequence).map_err(|e| {
            log::error!("listing sequences: {e}");
            DomainError::InvalidInput
        })?;
        rows.collect::<Result<Vec<_>, _>>().map_err(|e| {
            log::error!("scanning sequences: {e}");
            DomainError::InvalidInput
        })
    }

    fn get_by_prefix_and_year(
        &self,
        prefix: &str,
        year: i32,
    ) -> Result<InvoiceSequence, DomainError> {
        let conn = self.conn.lock().unwrap();
        conn.query_row(
            "SELECT id, prefix, next_number, year, format_pattern FROM invoice_sequences WHERE prefix = ?1 AND year = ?2 AND deleted_at IS NULL",
            params![prefix, year], scan_sequence,
        ).map_err(|e| match e {
            rusqlite::Error::QueryReturnedNoRows => DomainError::NotFound,
            _ => { log::error!("querying sequence by prefix/year: {e}"); DomainError::InvalidInput }
        })
    }

    fn count_invoices_by_sequence_id(&self, sequence_id: i64) -> Result<i64, DomainError> {
        let conn = self.conn.lock().unwrap();
        conn.query_row(
            "SELECT COUNT(*) FROM invoices WHERE sequence_id = ?1 AND deleted_at IS NULL",
            params![sequence_id],
            |row| row.get(0),
        )
        .map_err(|e| {
            log::error!("counting invoices by sequence: {e}");
            DomainError::InvalidInput
        })
    }

    fn max_used_number(&self, sequence_id: i64) -> Result<i64, DomainError> {
        let conn = self.conn.lock().unwrap();
        let result: Option<i64> = conn.query_row(
            "SELECT MAX(CAST(SUBSTR(invoice_number, LENGTH((SELECT prefix FROM invoice_sequences WHERE id = ?1)) + 5) AS INTEGER)) FROM invoices WHERE sequence_id = ?1 AND deleted_at IS NULL",
            params![sequence_id], |row| row.get(0),
        ).map_err(|e| { log::error!("max used number: {e}"); DomainError::InvalidInput })?;
        Ok(result.unwrap_or(0))
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::test_db::new_test_db;

    #[test]
    fn test_create_and_get() {
        let conn = new_test_db();
        let repo = SqliteSequenceRepo::new(conn);

        let mut seq = InvoiceSequence {
            id: 0,
            prefix: "FV".to_string(),
            next_number: 1,
            year: 2025,
            format_pattern: "{prefix}{year}{number:04d}".to_string(),
        };
        repo.create(&mut seq).unwrap();
        assert!(seq.id > 0);

        let fetched = repo.get_by_id(seq.id).unwrap();
        assert_eq!(fetched.prefix, "FV");
        assert_eq!(fetched.year, 2025);
    }

    #[test]
    fn test_list() {
        let conn = new_test_db();
        let repo = SqliteSequenceRepo::new(conn);

        let mut s1 = InvoiceSequence {
            id: 0,
            prefix: "FV".to_string(),
            next_number: 1,
            year: 2025,
            format_pattern: String::new(),
        };
        let mut s2 = InvoiceSequence {
            id: 0,
            prefix: "ZF".to_string(),
            next_number: 1,
            year: 2025,
            format_pattern: String::new(),
        };
        repo.create(&mut s1).unwrap();
        repo.create(&mut s2).unwrap();

        let list = repo.list().unwrap();
        assert_eq!(list.len(), 2);
    }

    #[test]
    fn test_get_by_prefix_and_year() {
        let conn = new_test_db();
        let repo = SqliteSequenceRepo::new(conn);

        let mut seq = InvoiceSequence {
            id: 0,
            prefix: "FV".to_string(),
            next_number: 1,
            year: 2025,
            format_pattern: String::new(),
        };
        repo.create(&mut seq).unwrap();

        let found = repo.get_by_prefix_and_year("FV", 2025).unwrap();
        assert_eq!(found.id, seq.id);

        let not_found = repo.get_by_prefix_and_year("FV", 2024);
        assert!(matches!(not_found, Err(DomainError::NotFound)));
    }
}
